dashboard/src/reducers/charts.test.ts

```diff
@@ -102,61 +105,81 @@ describe("chartReducer", () => {
   it("single requestCharts with query", () => {
     const state = chartsReducer(undefined, {
       type: getType(actions.charts.requestCharts) as any,
-      payload: "query",
+      payload: 1,
+      meta: "query",
     });
     expect(state).toEqual({
       ...initialState,
       isFetching: true,
     });
   });
 
-  it("single receiveCharts without query", () => {
+  it("single receiveCharts no finished", () => {
     const state = chartsReducer(undefined, {
       type: getType(actions.charts.receiveCharts) as any,
-      payload: [chartItem],
+      payload: { items: [chartItem], page: 1 } as IReceiveChartsActionPayload,
+      meta: false,
     });
     expect(state).toEqual({
       ...initialState,
       isFetching: false,
+      hasFinished: false,
       items: [chartItem],
+      page: 2,
+      records: new Map<number, boolean>().set(1, true).set(2, false),
     });
   });
 
-  it("single receiveCharts with query", () => {
+  it("single receiveCharts finished", () => {
     const state = chartsReducer(undefined, {
       type: getType(actions.charts.receiveCharts) as any,
-      payload: [chartItem],
+      payload: { items: [chartItem], page: 1 } as IReceiveChartsActionPayload,
+      meta: true,
     });
     expect(state).toEqual({
       ...initialState,
       isFetching: false,
+      hasFinished: true,
       items: [chartItem],
+      page: 1,
+      records: new Map<number, boolean>().set(1, true).set(2, false),
     });
   });
 
-  it("two mixed receiveCharts should override items", () => {
+  it("two receiveCharts should add items", () => {
     const state1 = chartsReducer(undefined, {
       type: getType(actions.charts.receiveCharts) as any,
-      payload: [chartItem],
+      payload: { items: [chartItem], page: 1 } as IReceiveChartsActionPayload,
+      meta: false,
     });
     const state2 = chartsReducer(state1, {
       type: getType(actions.charts.receiveCharts) as any,
-      payload: [],
+      payload: { items: [], page: 2 } as IReceiveChartsActionPayload,
+      meta: true,
     });
     expect(state2).toEqual({
       ...initialState,
       isFetching: false,
-      items: [],
+      hasFinished: true,
+      items: [chartItem],
+      page: 2,
+      records: new Map<number, boolean>()
+        .set(1, true)
+        .set(2, true)
+        .set(3, false),
     });
   });
 
```

Comment:
I'd add a couple more of tests, one to check that receiving two pages of charts includes all the charts as `items` and another checking that if you add the same page twice the items are not duplicated.